### ISSUE TITLE
Inline JS and CSS and Content Security Policy

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#1142]
Currently, the reliance on inline JavaScript and inline CSS makes it near impossible to create a secure [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) for Textpattern.

Sure, you can set strict rules in your root `.htaccess` file that disallow inline JS and CSS for the site itself but you then need to create another `.htaccess` in your `textpattern` subdirectory with `'unsafe-inline'` directives set to get Textpattern's control panel to work. That takes away a large chunk of the security provided by CSP.

If possible we need to move core JavaScript to external files and use classnames instead of inline CSS `style` attributes.

Obviously plugin authors are a separate issue, but at least we can then say the core can be covered by CSP.

**Comment by Bloke** [comment]
Inline JS is something we should be 90% of the way to designing out. Oleg's done some great work in `textpattern.js` this release to let us use the Routing capability in a more flexible manner. Plus there's the whole "rendering stuff in a View" (`Admin/Table.php` for starters) that'll be coming online, so panel generation and visualization will be more centralized. That should cut down a lot of variation and allow us to leverage reuse more.

There was a bit of inline JS to handle the search widget, but that may have gone now, not sure. There's a line or two of inline CSS in `txp_diag.php` that we should ideally get rid of now. Can't really do much about the public tags that take a `style` attribute and therefore, by inference, the tag builder. Beyond that I haven't paid much attention to the specifics. Are there places in particular that you see inline CSS/JS that we need to stamp out?

**Comment by philwareham** [comment]
In the `<head>` the inline `<script>` that defines `var textpattern` is a biggie. As is the `<script>` in the `<footer>`.

CSS is mainly stuff like `style="display:none"` which could and should be replaced with a suitable `class` and so on. CSS isn't really as crucial as JS as it's less of a security risk but if we can shore it up we should.

**Comment by bloatware** [comment]
The main source of inline scripts are imo the dynamic parameters/preferences/strings that we need to pass to static `textpattern.js`. Dunno what is the best solution here. We certainly could serve JS files via PHP, but it would mean extra processing time.

More globally, `textpattern.js` is becoming a mess (mea culpa), so something must be done in 4.8 anyway.

**Comment by philwareham** [comment]
I'll mop up some of the easy CSS ones.

**Comment by Bloke** [comment]
Yeah, the `display:none` in `txp_diag.php` and perhaps elsewhere should be easy to stamp out.

I'd forgotten about the inline `textpattern` global. Yes that is a problem as you say, passing stuff from server to client. Would be great to figure out how to do that cleanly - can we introduce 'head' as a special `Textpattern.Route` or something?

As Oleg says, we also have a lot of cruft in `textpattern.js` - old functions, even obsolete or superseded stuff that we can either remove, do better, or modularise.

Speaking of which, is it possible to modularise the various JS components? Like we do with SASS/SCSS, and "compile" them into one minified JS file? Or is that too much trouble?

**Comment by philwareham** [comment]
Sure, we can easily modulise the JavaScript and use Uglify to concatenate. Let’s do that in 4.8.

**Comment by petecooper** [comment]
Is there an ideal/preferred/current CSP doc for Textpattern? That seems like a logical step as part of ironing out some snags.

**Comment by philwareham** [comment]
I've removed all the inline CSS I can now. But due to the extensive use of `style="display:none"` in jQuery and jQuery UI, I think we'll just have to document that you can add this hash into your CSP `style-src` directive:

    'sha256-aqNNdDLnnrDOnTNdkJpYlAxKVJtLt9CtFLklmInuUAE='

Maybe we can document a recommended set of CSP settings for Textpattern, if we can get the inline JS down to a manageable amount?

**Comment by philwareham** [comment]
I've gone round and round on using a server-generated dynamic random nonce value in the CSP `script-src` directive and then also reusing that same value on a `nonce=""` HTML attributed in any inline JS created by our `script_js()` PHP function, and I can't find a satisfactory solution.

If anyone else has any ideas then please comment, otherwise I guess we will just have to live with `'unsafe-inline` CSP value on `script-src` - which isn't secure at all - or someone with PHP and JS skills may be able to take all our inline JS and somehow make it external? Not sure that is feasible.

**Comment by Bloke** [comment]
When I was developing ext_recaptcha, I hit this issue. Google prefer the inline js to be made with a nonce value that's reflected in the page header.

I got round it by using a tiny plugin (like, 2 lines: hardly even worth making a plugin) called smd_token that generates a random nonce on each page load. That value can be stashed in a `<txp:variable>` and plugged into the recaptcha plugin as a `nonce` attribute which renders the value in the `<script>` tag.

A corresponding header tag adds the value to the page headers:

`<txp:header name="Content-Security-Policy" value='script-src "nonce-<txp:variable name="csp-nonce" />"' replace="-1" />`

Job done. I can share this approach if it would help?

**Comment by philwareham** [comment]
> Job done. I can share this approach if it would help?

As far as I can tell, we'd have to declare the whole CSP header again, not just the `script-src` and its value (as long as headers set via Textpattern take precedence over headers set by the server). Whilst we could do that for our official sites without too much fuss, I was looking for a solution that would feasible for all users that want to use CSP.

Maybe if your 2 lines of code that creates the nonce was available in core, it could then be available to end users to use as they wish, including for a page header (a random nonce generated at page load could be useful for other purposes too).

**Comment by Bloke** [comment]
> Maybe if your 2 lines of code that creates the nonce was available in core, it could then be available to end users to use as they wish

Ready? Brace yourself... here's the entirety of the plugin, bar all the fluff and tag registration and attribute handling:

``` php
$token = Txp::get('\Textpattern\Password\Random')->generate(16);
```

:)

That's it. Yes the plugin was for public side as a convenience, but the principle still applies to the admin-side.

Plug that token value in:

```
echo '<script nonce="'.$token.'" src="https://example.org/some/script.js"></script>';
header('Content-Security-Policy: script-src "nonce-'.$token.'"');
```

**Comment by philwareham** [comment]
Any chance of you getting this into 4.8.2 branch? I can then test against a server-set CSP header and see if that works.

**Comment by Bloke** [comment]
We can try. Shall we do it with `textpattern.js` for starters just to test it?

**Comment by bloatware** [comment]
> headers set via Textpattern take precedence over headers set by the server

probably not with `replace="-1"` attribute, though I'm not sure.

**Comment by philwareham** [comment]
> We can try. Shall we do it with textpattern.js for starters just to test it?

OK, let's do that 😀

> One fresh token value per script should do the trick.

I think the same nonce value can be used throughout the page, on multiple script tags, as long as it randomly changes at each page load.

**Comment by Bloke** [comment]
> OK, let's do that 😀

Actually, we _could_do it in `script_js()`. Maybe pass in a TEXTPATTERN_SCRIPT_NONCE flag which adds a random value to the returned tag and adds the header somehow? Not sure if that'd work.

However, that would generate a new value for each page load unless we cached the value and used the same one each time.

> I think the same nonce value can be used throughout the page, on multiple script tags, as long as it randomly changes at each page load.

I didn't know that, but it makes sense I guess. Less gumph in the header too. In which case, caching the token is the way to go.

**Comment by Bloke** [comment]
Alternatively, instead of passing in a flag, do we want this to happen all the time? Are there any instances - plugins or otherwise, admin-side or front-end - where we wouldn't want this behaviour in the `script_js()` function?

**Comment by philwareham** [comment]
Having the nonce on all instances is fine, even if the script is an external one. In fact, that can be more secure than using `'self'` on `script-src` and it also makes plugins with inline JS secure (as long as they use `script_js()` for scripts).

**Comment by Bloke** [comment]
Hmm, this looks like it can't be opt-in. It's either on or not. The addition of the nonce value in the header seems to imply in my limited testing so far that _all_ scripts (including inline ones) need a nonce or they all get rejected.

However, according to [the W3C](https://www.w3.org/TR/CSP2/#nonce_value):

> Note: Using a nonce to whitelist inline script or style is less secure than not using a nonce, as nonces override the restrictions in the directive in which they are present. An attacker who can gain access to the nonce can execute whatever script they like, whenever they like. That said, nonces provide a substantial improvement over 'unsafe-inline' when layering a content security policy on top of old code. When considering 'unsafe-inline', authors are encouraged to consider nonces (or hashes) instead.

So where exactly does that leave us?

**Comment by philwareham** [comment]
Use 'self' in addition to the nonce on the header rule.

**Comment by Bloke** [comment]
> Use 'self' in addition to the nonce on the header rule.

D'oh of course. Well that fixes _most_ of the issues. But there are a few remaining:

1. As soon as the `header()` is added, one of the inline scripts fails. I've no idea which one as they all seem to have the `nonce-` value added in the source. And clicking in the inspector doesn't take me to the line that's the problem. The only scrip tas that don't have the nonce are the ones added by the theme. Which brings me to...
2. ... our `script_js()`isn't wired up to be able to accept other attributes such as `defer` so it can't be used at the moment in the theme. Without `defer`, hive breaks. We'd need to find a way to pass that and (potentially) other attributes in.

One way to fix point 2, is to not do it now and do this in 4.9 when we can add a new UI class to handle script output. That makes it a breeze to add any attributes, as the classes take care of it all for us.

Thoughts? Any idea which inline script might be the one throwing this out?

**Comment by philwareham** [comment]
> One way to fix point 2, is to not do it now and do this in 4.9

Of course, the issue is targeted at 4.9 anyway. In the meantime I can leave our server CSP with `'unsafe-inline'` as it has been for last couple of years, or maybe Pete can do a specific CSP block just for the 'textpattern' directories in Nginx config.

> Thoughts? Any idea which inline script might be the one throwing this out?

Not without seeing the code running, no, sorry.

**Comment by Bloke** [comment]
If you want to try the no-frills version, replace your `script_js()` function in lib/txplib_html.php with this:

``` php
function script_js($js, $flags = '', $route = array())
{
    static $store = '', $token = null;
    global $event, $step;

    if ($token === null) {
        $token = base64_encode(Txp::get('\Textpattern\Password\Random')->generate(PASSWORD_LENGTH));
        header("Content-Security-Policy: script-src 'self' 'nonce-$token'", false);
    }

    $targetEvent = empty($route[0]) ? null : (is_array($route[0]) ? $route[0] : do_list_unique($route[0]));
    $targetStep = empty($route[1]) ? null : (is_array($route[1]) ? $route[1] : do_list_unique($route[1]));

    if (($targetEvent === null || in_array($event, $targetEvent)) && ($targetStep === null || in_array($step, $targetStep))) {
        if (is_int($flags)) {
            if ($flags & TEXTPATTERN_SCRIPT_URL) {
                if ($flags & TEXTPATTERN_SCRIPT_ATTACH_VERSION && strpos(txp_version, '-dev') === false) {
                    $ext = pathinfo($js, PATHINFO_EXTENSION);

                    if ($ext) {
                        $js = substr($js, 0, (strlen($ext) + 1) * -1);
                        $ext = '.'.$ext;
                    }

                    $js .= '.v'.txp_version.$ext;
                }

                return n.tag(null, 'script', array('src' => $js, 'nonce' => $token));
            }
        }

        $js = preg_replace('#<(/?)(script)#i', '\\x3c$1$2', $js);

        if (is_bool($flags)) {
            if (!$flags) {
                $store .= n.$js;

                return;
            } else {
                $js = $store.n.$js;
                $store = '';
            }
        }

        $js = trim($js);
        $out = $js ? n.tag(n.$js.n, 'script', array('nonce' => $token)) : '';

        if ($flags && $flags !== true) {
            $out .= n.tag(n.trim($flags).n, 'noscript');
        }

        return $out;
    }

    return '';
}
```

Then you'll see what I see.

**Comment by philwareham** [comment]
> Hmmm, does the <noscript> tag need a nonce?

No that HTML tag can't tag that attribute.

**Comment by Bloke** [comment]
So if it's not noscript then I can't figure which one it is that's borking this. I presume that textpattern.js is generating some script response without a nonce which is why Ajax saves are failing. No ideas how to fix that.

**Comment by petecooper** [comment]
>maybe Pete can do a specific CSP block just for the 'textpattern' directories in Nginx config.

Yeah, he can do that.

**Comment by philwareham** [comment]
Leave it with me and I'll try the code out tomorrow. Cheers all.

**Comment by philwareham** [comment]
Further investigation on this for reference:

1. We do already actually have some CSP stuff built into Textpattern 3 years ago by @makss - I didn't even remember this happening but it's here:

https://github.com/textpattern/textpattern/commit/f004d363afc8ce271d058220dacd5a2561e74eab

That means we can probably amend Stef's code a bit to use what is there to some extent, and move the random nonce generator to a place where is can be utilised by a user in the config file (instead of setting the header automatically in the `script_js()`).

2. Setting a new CSP header in PHP **does keep rules** from the server conf/virtualhost/.htaccess in place (any least, it does on Apache, anyway) **but** if the upstream CSP directive is tighter than the downstream directive then the looser directive is ignored. Basically, for security, you can't loosen existing CSP, only tighten it - which makes sense.

i.e. If you had this in conf:

    Content-Security-Policy script-src 'self';

You can't put this in a PHP file to relax it:

    header("Content-Security-Policy: script-src 'self' 'nonce-$token'", false);

Because self + nonce is less secure than the upstream self.

3. To allow less secure CSP in the Textpattern subdirectory you'd need to do something like this in .htaccess or equivalent:

    Header unset Content-Security-Policy

Which will then allow you to set a new CSP policy from scratch.

TLDR: we should allow the nonce random string, that is set on `<script>` tags to be used in the config file as part of a user-defined CSP. Nothing should be done automatically by the core itself regarding changing CSP headers.

The constant we do have as mentioned in point 1 above is OK to leave in place.

**Comment by Bloke** [comment]
> move the random nonce generator to a place where is can be utilised by a user in the config file (instead of setting the header automatically in the `script_js()`).

Hmmm, I'm not sure how we'd do that. config.php is pretty much [the very first thing that's loaded](https://github.com/textpattern/textpattern/blob/dev/textpattern/index.php#L36) because it's vital for database credentials. So the only way to set a nonce that users can add to config.php would be do it _in config.php_ itself, or in the line before it's included in index.php.

I'm not really sure what that solves though. Adding it to our default header in constants.php is surely better? Then at least, users can override it in config.php if they want with their own implementation, otherwise they get our CSP rules.

BUT we can't do that using the core class libraries because at the very moment constants.php is loaded, the class loader hasn't been initialized. And we need the constants (and txplib_misc.php and txplib_admin.php) to be loaded _before_ the class loader is instantiated.

Chicken and egg if we try and do it early like that, unfortunately. The only way to do it would be to _not_ use our class library and hardcode the nonce generator in constants.php, which is not something I'd like to do.

Hmmmmm.

**Comment by Bloke** [comment]
> self + nonce is less secure than the upstream self only.

Really?! Wow. I would have thought that adding a random token was _more secure_ than relying on 'self' but maybe that rule has some in-built security measures that would be undone by adding the later rule.

That doesn't explain why it "broke" when I did this:

``` php
header("Content-Security-Policy: script-src 'nonce-$token'", false);
```

And may explain why this worked:

``` php
header("Content-Security-Policy: script-src 'self' 'nonce-$token'", false);
```

Because the rule was effectively being kicked out by the policy parser for being too relaxed? Uhhh, maybe?

**Comment by philwareham** [comment]
> Really?! Wow. I would have thought that adding a random token was more secure than relying on 'self'

'random token' alone would be, however 'self **+** random token' is not as secure as 'self' alone, if you see what I mean.

I have to say, this CSP stuff is *stupidly* hard to implement. I wish I'd never gone down this rabbit hole in the first place!

**Comment by Bloke** [comment]
> 'random token' alone would be, however 'self **+** random token' is not as secure as 'self' alone, if you see what I mean.

Gotcha. That's what I figured after I read your overview above. And explains my findings yesterday, even though I didn't understand why it was "broken" at the time.

So yeah, as you say, stupidly difficult to implement. If we append a rule (without 'self' in it) then _all_ scripts on the page MUST use the nonce. That includes plugins, themes, and everything. Any that don't will be blocked.

That's fine if all plugins and themes use `script_js()` because they'd get the nonce for free. But anyone that hard-codes `<script>` tags in their admin-side code are out of luck. Instant plugin breakage.

Same issue, presumably, if we extend this idea to CSS; any hard-coded `<style>` tags without a nonce would be rejected. And in that case, I don't recall us having a centralised API function for generating style markup, so we'd need to build one and encourage everyone to use it.

Minefield, as you say.

So going back to the original issue - and excuse the ignorant question - if 'self' is already in place and only permits scripts from the same domain to be loaded, what does overriding it with a nonce add? Or are we trying to fix this so we can tag third party script requests with a unique token so they can't be tampered with by man-in-the-middle shenanigans?

**Comment by philwareham** [comment]
OK, so here's where I'm at with this, if a user wants to use CSP:

1. User has nice secure CSP in their root.
2. User can unset their root CSP header for the Textpattern directory specifically.
3. User then defines their CSP in the config.php file like so:

```
$headerCsp =
    "base-uri 'none';".
    "block-all-mixed-content;".
    "connect-src 'self';".
    "default-src 'none';".
    "font-src 'self';".
    "form-action 'self';".
    "frame-ancestors 'self';".
    "img-src data: 'self';".
    "media-src 'self';".
    "script-src 'self' 'unsafe-inline';".
    "style-src 'self' 'unsafe-inline';".
    "style-src-attr 'self';".
    "style-src-elem 'unsafe-inline';".
    "worker-src 'none'";

define('CONTENT_SECURITY_POLICY', $headerCsp);
```

That gives us fairly good CSP so far, yes we have `'unsafe-inline'` on scripts and styles but the style one is then tightened somewhat in CSP level 3 compliant browsers by using `script-src-attr`, `script-src-elem`, `style-src-attr` and `style-src-elem` directives and using the nonce values on the `script-src-elem` and `style-src-elem` *if we can get it working*.

So the additions I think would be:

```
    "style-src-attr 'sha256-whatever';". (or we just 'unsafe-inline' this one as the threat would be negligible)
    "style-src-elem 'nonce-$whatever';".
    "style-src-attr 'sha256-whatever';". (or we just 'unsafe-inline' this one as the threat would be negligible)
    "style-src-elem 'nonce-$whatever';".
```

**Comment by Bloke** [comment]
Nice work! Should we include that as our default CSP policy in constants.php? Or are we keeping it permissive there?

**Comment by philwareham** [comment]
We can’t really use this in constants - there’s too many assumptions made. Any plugin that links to a third party service or CDN or whatever would break if the above rules were enforced.

Bette to either comment this in config or document it somewhere official.

I’d still like to pursue the random nonce though - can’t we generate that with php in the config then use in the script function (if defined)? Again commented out or documented somewhere.

I think have a CMS that works with CSP well is a good win. Don’t think many CMSes can achieve that.

If that nonce could also be used on front side by core tags, even better.

**Comment by petecooper** [comment]
>I think having a CMS that works with CSP well is a good win.

Big +1 to this. Even providing documentation to show we have _awareness_ of stuff outside of the 'just write' approach will be invaluable.

**Comment by Bloke** [comment]
> I’d still like to pursue the random nonce though - can’t we generate that with php in the config then use in the script function (if defined)?

Yes we can. We'd just have to replicate some/most of the functionality of our [Random->generate() method](https://github.com/textpattern/textpattern/blob/dev/textpattern/vendors/Textpattern/Password/Random.php#L69) in constants.php which, to me, is a shame. I'd rather call that method directly as it's just one line of code then. But as I mentioned above, at the time constants.php is loaded, we don't have access to our class library. By the time it's loaded, it's too late. constants.php has been run, config.php has been read and we're on our way to authenticating and rendering the page.

We can't swap the order of things here because the class library relies on constants being defined in, well, constants.php - so I'm pretty sure that [has to come first](https://github.com/textpattern/textpattern/blob/dev/textpattern/index.php#L51). See [$loader->register();](https://github.com/textpattern/textpattern/blob/dev/textpattern/index.php#L57) a few lines later on.

So we either duplicate the random generator code, or find a shorter way to do what that method does and implement it in constants.php. Or introduce some well-known pre-config file that gets included (and hope that doesn't pose a security risk).

While I agree that having CSP in place would be a boon, I'd like to explore ways we could expose this in a more API-ish way if possible. Enable people to opt-in and modify the CSP via a callback perhaps. Even if that callback isn't like our regular plugins but simply reads in a particular set of variables from config.php, or we defer output of the CSP header until we've merged in the contents of some user configuration so we only output one header.

I don't exactly know the best way to handle it right now.

Let's figure out the best pathway for us to offer baseline CSP filters out of the box which can then be augmented/replaced somehow by user config prior to the header being output, and then allow those tokens to be used on the admin-side (and potentially public side if we drop its value in the `textpattern` JavaScript object).

**Comment by philwareham** [comment]
I don’t think we should be setting any of the users CSP (even the one directive we go set currently I’d personally drop) - rather, give them the tools and method to do their CSP if they wish to see in a A+ rated way (my example above without nonce is A rated but still can be compromised due to unsafe inline allowance).

We need a nonce set before the config.php is read, and also so it can be used in functions.

**Comment by Bloke** [comment]
We can take out the one that's there in constants.php. That way it's a blank slate.

I don't see that we need to generate the nonce before the config is read. We need the nonce generated before the `header()` is sent. So it can be merged with whatever other rules they decide they want.

So we can play it a few ways. Firstly, allow this in config.php:

``` php
$headerCsp =
    "base-uri 'none';".
    "block-all-mixed-content;".
    "connect-src 'self';".
    "default-src 'none';".
    "font-src 'self';".
    "form-action 'self';".
    "frame-ancestors 'self';".
    "img-src data: 'self';".
    "media-src 'self';".
    "script-src 'self' 'unsafe-inline';".
    "style-src 'self' 'unsafe-inline';".
    "style-src-attr 'self';".
    "style-src-elem 'unsafe-inline';".
    "worker-src 'none'".
    "script-src-attr 'sha256-somehash';".
    "script-src-elem 'nonce-{TEXTPATTERN_CSP_NONCE}';".
    "style-src-attr 'sha384-somehash';".
    "style-src-elem 'nonce-{TEXTPATTERN_CSP_NONCE}';";

define('CONTENT_SECURITY_POLICY', $headerCsp);
```
Later, just as we come to output the headers, if we find `CONTENT_SECURITY_POLICY`, we generate a global nonce at that moment and `str_replace()` it in the string, then output it.

The trick here would be to determine if a nonce has been specified. Then, when it comes time to output scripts or styles, we know to add that attribute to the relevant script/style tags and to make it available as a global to Txp tags/plugins, long term, perhaps.

I'm not sure of the practicality of using the inline hashes. You need to compute those using one of the pre-approved algorithms (or use the sha256 that the browser suggests in the violation report) and add one per script, right? Whereas a single nonce can be reused across the page, for all scripts and styles. But we can support them if people can be arsed to add them all to their config.php.

The second approach is a variation on the above. But instead of using a replacement string, you specify your bare-bones, fixed header elements and then define a second flag like `define('TEXTPATTERN_CSP_NONCE', true);` which we can then detect at the moment we output the header. At that point, we generate the nonce and stash it, add it to the header string already hard-coded, and output it. Then any script tags that are generated through our internal functions `script_js()`and the (currently fictitious) `style_tag()` (or something) get the `nonce` attribute added.

I'm easy with either approach. And I think both are workable.

**Comment by philwareham** [comment]
Ah, I see, I was mistakenly thinking the `TEXTPATTERN_CSP_NONCE` had to already be defined in order to use - that's good then.

> I'm not sure of the practicality of using the inline hashes.

This would be for inline style added in a HTML `style` attribute (`style-src-attr`) as opposed to a `<script>` tag (`style-src-elem`) where the HTML element can't accept a `nonce` attribute.

Although, I'm having a bit of trouble getting hashes working for inline styles via `style` attributes in Chrome (the only browser that supports those new CSP directives so far). Doesn't help that the MDN documentation for these is currently incomplete, and no examples given, so I may be doing something wring but I don't know what. Using the hashes that Chrome suggest in the console doesn't;'t work.I think I'll use `unsafe-inline` for `style-src-attr` for now.

**Comment by philwareham** [comment]
> Nothing we can do about the hard-coded ones or those from plugins. They'll need updating or they'll get blocked.

That's why I'm proposing we just supply the method, not enforce it as standard. It's up to the user to sort out their CSP and if they use plugins then it's down to them to either find a workaround, use a less restrictive CSP, or ask the plugin author to update the plugin to comply.

**Comment by Bloke** [comment]
> I'm having a bit of trouble getting hashes working for inline styles via `style` attributes

From what I read last night, it seems you just add a series of rows like this:
```
Content-Security-Policy: style-src-attr 'sha256-B2yPHKaXnvFWtRChIbabYmUBFZdVfKKXHbWtWidDVF8='
```
(or `style-src-elem`: I'm not sure of the difference) and then just use bare `<style>` tags in the markup. Quite how the browser figures out which is which I'm not sure. Maybe it just computes the hashes of all style blocks and makes a list then checks they all match?

**Comment by philwareham** [comment]
Yes, that's the theory - however I have tried the following style hash countless times:

    'sha256-aqNNdDLnnrDOnTNdkJpYlAxKVJtLt9CtFLklmInuUAE='

Which is a hash of:

    display:none

And it never works.

I've used hashes of scripts in sites before as part of their CSP and not had any problems. Just don't know why it isn't working for style.

**Comment by Bloke** [comment]
Hmm, according to [this generator](https://passwordsgenerator.net/sha256-hash-generator/) and then running the resultant sha256 string through [this base64 encoder](https://www.base64encode.org/) the hash of `display:none` without any leading and trailing spaces is:

`NkFBMzRENzQzMkU3OUVCMENFOUQzMzVEOTA5QTU4OTQwQzRBNTQ5QjRCQjdEMEFEMTRCOTI1OTg4OUVFNTAwMQ==`

So `<style>display:none</style>` (no leading/trailing spaces inside the style tags) ought to be represented by this CSP rule:

```
Content-Security-Policy: style-src-attr 'sha256-NkFBMzRENzQzMkU3OUVCMENFOUQzMzVEOTA5QTU4OTQwQzRBNTQ5QjRCQjdEMEFEMTRCOTI1OTg4OUVFNTAwMQ=='
```

But I don't know why the one created by that generator is so long compared with the ones thrown onto the browser console.

**Comment by petecooper** [comment]
@philwareham - is this of any relevance? https://stackoverflow.com/questions/52724956/why-doesnt-chrome-respect-my-content-security-policy-hashes/52795613#52795613

**Comment by philwareham** [comment]
Actually, I think I need to add `unsafe-hashes` before the actual hash: https://content-security-policy.com/unsafe-hashes/

Source: https://bugs.chromium.org/p/chromium/issues/detail?id=546106

I'll see if that fixes my problems - because jQuery `.hide()` and `.show()` are, I guess, event-based they are probably breaking the CSP rules without that `unsafe-hashes` rule.

Will let you know.

**Comment by philwareham** [comment]
> is this of any relevance?

In some ways, yes, although it is outdated info.

**Comment by philwareham** [comment]
Hooray, adding `unsafe-hashes` solves the issue in Chrome and is moire secure than leaving it as `unsafe-inline`.

So, to summarise - state a baseline CSP rule:

    style-src 'self' 'unsafe-inline';

Tighten that with further rules for CSP level 3 complaint browsers:

    style-src-attr 'unsafe-hashes' 'sha256-aqNNdDLnnrDOnTNdkJpYlAxKVJtLt9CtFLklmInuUAE=';
    style-src-elem 'nonce-OurNonceValue';

I'd have to go through the codebase and a vanilla local Textpattern install to see if more hashes need to be added to `style-src-attr` - but it's progress.

@bloke can you please, when you have spare time, do the changes to nonce generator in the `csp-testing` branch? I've put a commented out CSP in the `config-list.php` file of that.

**Comment by Bloke** [comment]
Nice! Yeah, I'll do that over the weekend.

Got any preference on which of the approaches I outlined earlier would be best?

**Comment by philwareham** [comment]
Probably approach 1, seems more concise?

**Comment by Bloke** [comment]
First one works for me. Will we automatically append _both_ script and style nonce headers then? Since there's no way in approach one to specify which one you're targeting.

**Comment by Bloke** [comment]
Approach 1 it is.

**Comment by Bloke** [comment]
Okay, please test 6f1aa85e. It adds support for `script-src` level checking. If you include one of the `script-src` rules in your CSP header AND it starts with `nonce` then it triggers `script_js()` to add the `nonce `attribute. Otherwise it remains silent.

Note that it _only_ checks `script-src` so if you add style-src rules with a nonce attribute, the `script_js()` isn't affected. We should probably introduce some sort of style rendering function  that behaves in a similar manner so people can benefit from style-src rules without having to add nonces by hand.

Interesting side-read, btw on [nonce resuse](https://w3c.github.io/webappsec-csp/#security-nonces). Of particular note is the nonce retargeting section. Should we add a `<base>` tag to our admin-side pages as a matter of course? That would mean a `base-uri` directive is unnecessary and adds a layer of protection by default.

**Comment by philwareham** [comment]
Looking good - I like that it only generates nonce attributes if it detects a CSP setup in config.php, very nice!

It would be good to get the flexibility of having the 'nonce' wherever order in the CSP directives, as you say. So if a way around this issue could be found that would be great. It's not a deal-breaker if not (but might create support queries from users).

I need to update the admin themes so that they use the `script_js` function too and inherit the nonce - which I'll do ASAP.

Style tags will also need the nonce, can that be done with the tools we have now or does a new function need to be created?

> Interesting side-read, btw on nonce reuse. Of particular note is the nonce retargeting section.

By us explicitly stating...

    base-uri 'none';

...in the CSP (disallowing any use of `<base>` tag), that threat is negated. Something to be aware of though, sure.

**Comment by Bloke** [comment]
> I need to update the admin themes so that they use the `script_js` function too and inherit the nonce - which I'll do ASAP.

You might have fun here. The `script_js()` function can't currently add `defer` so the admin side breaks. Options:

* Use `\Textpattern\UI\Tag` to create your own script tag.
* Wait until someone adds a `UI\Script` wrapper for the script tag.

> Style tags will also need the nonce, can that be done with the tools we have now or does a new function need to be created?

New function. Again, I'm inclined to not make a function and just use the UI class here since it's way more flexible.

**Comment by philwareham** [comment]
Sure, I can `defer` this issue until you are ready with the UI widgets. Did you see what I did there? 👏👏👏

**Comment by Bloke** [comment]
I've modified the regex so please try f54fd022 and see if you can now insert extra rules either side of the script-src directives. I told it to match one or more of those rules and to ensure the final one ends with a semicolon so _hopefully_ it'll treat each rule as a separate item. Not pretty so if anyone has any improvements, please do so.

**Comment by philwareham** [comment]
Yep, that works in testing. So now to complete this issue we need UI widgets for script and style tags. No rush though.

**Comment by Bloke** [comment]
Cool, glad it works. I'll add the UI element wrappers, hopefully tonight.

**Comment by philwareham** [comment]
I've hopefully fixed certain AJAX from our textpattern.js being blocked in CSP level 3 browsers now (e.g. Chrome).

Final item is the style UI widget. Then I think we can merge this branch into dev and close the issue. Feature will need documenting somewhere for user reference prior to 4.9 release (I've left a placeholder for that link in the CSP-related comment in config.php too).

**Comment by Bloke** [comment]
Cool. Once I've retrofitted `script_js()` to work with the new class (nearly done: one niggle to iron out later) then I'll do a similar 'Style' class.

**Comment by philwareham** [comment]
Just a gentle bump of this - once the `style` widget is done, I can then close this issue, merge branch and then put my weight behind the image refactoring work. Cheers!

**Comment by Bloke** [comment]
The Style tag is a little more complex than I'd hoped. The problem is that there are _two_ types of tag that can be generated. An inline `<style>` tag (if you pass in/append just content) or a `<link>` tag (if you pass in a URL).

At a superficial level, it works just like the '<script>' tag but the hard part is we don't know (when you first instantiate this class), which one you're going to use until _after_ you've started adding content to it. Just means we have to defer construction of the type of tag until the `render()` stage.

Working on it...

**Comment by philwareham** [comment]
Can’t we have two tags? One for inline styles and one for linked styles? Dunno, sounds fine to me to have two.

**Comment by Bloke** [comment]
We could, yeah. That gives us the ability to use the Link tag for other purposes. Just seems a bit overkill, since Link is so vast and generic in scope, you might as well just use the base Tag instead:

``` php
echo new Textpattern\UI\Tag('link')
   ->setAtt('rel', 'canonical')
   ->setAtt('href', 'https://some/link/endpoint');
echo new Textpattern\UI\Tag('link')
   ->setAtts(array(
      'rel' => 'stylesheet',
      'media' => 'screen',
   )
   ->setAtt('href', 'https://some/stylesheet.css');
...
```

Doesn't seem much point writing a generic link tag and then _only_ wiring up the CSP stuff if you happen to add `rel="stylesheet"`. Unless you can use CSP on other types of content that the `<link>` tag can be used for, such as icons and prev/next links and so forth?

In fact, can you even use `nonce`' on a `<link>`? If not then it becomes easy and we just use the Style class for actual `<style>` tags. Interestingly, MDN claims [nonce isn't a valid attribute on links](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link) but [html.am says it is](https://www.html.am/tags/html-link-tag.cfm). Two men say they're Jesus. One of them must be wrong.

**Comment by philwareham** [comment]
Link can have a nonce according to WhatWG https://github.com/whatwg/html/commit/27f03238c9a6a385879e51bd49c30453ce0d0b48

Although I’d only think CSP directives would be for stylesheets though.

**Comment by Bloke** [comment]
As far as I'm concerned, f15cd8da finishes the CSP work by introducing a Style tag that can be used to _either_ render CSP_aware inline `<style>` tags OR `<link rel="stylesheet">` tags. Switch usage by either using `->setSource()` (for links) or `->setContent()` (for styles). 

Similar to the Script tag, the latter method can also be called multiple times with `false` as the second parameter to append content, allowing you to build up concatenated style rule sets. You mark them as 'finished' ("output now") by supplying `true` as the second argument to flush the buffer and start a fresh `<style>` tag.

You can add any other attributes as normal, though it will 'force' the `rel="stylesheet"` if you `->setSource()`.

You can also target style tags only on specific panels (events or steps) by using `->setRoute()`. First param is a comma-separated list of events, second param is a comma-separated list of steps upon which the style tag is to appear. Great for plugins to be able to target only their own panels without having to check by hand (as we currently do).

In c6b2b83, I've introduced this new class to the three core admin themes so please test them with CSP rules and check they behave.

**Comment by philwareham** [comment]
I've just checked the latest commit with a tight CSP (based on the ones in the config-dist.php file we've defined) and all seems to be working as intended.

So I'm happy to merge this branch into dev if you are? Cheers!

**Comment by philwareham** [comment]
Regarding those other `link` types - we don't use them and I don't foresee us using them, so no worries there.

**Comment by Bloke** [comment]
Yeah, merge away into dev. One more branch pruned!

**Comment by philwareham** [comment]
Done and merged! Thanks everyone! 👏👏👏

**Comment by philwareham** [comment]
Wonder if we can squeeze this into 4.8.3 - don’t think any of the additions are breaking things if we selectively add the files that have changed. Thoughts?

**Comment by petecooper** [comment]
I'm almost getting ready to mention the usual Textpattern patch feature creep…so I'm a tentative +1 for getting it in a) if you know it's working and b) it doesn't delay 4.8.3 any further.

**Comment by petecooper** [comment]
Having said this ^^^ I know it's satisfying watching meaningful stuff ship, so perhaps it's a +1 from me and not a tentative one.

**Comment by Bloke** [comment]
Sadly, the only way this can make it in 4.8.3 is if the UI library comes with it. Without that, we can't patch `script_js()` and don't have the ability to offer style-based CSP.

**Comment by philwareham** [comment]
I thought the ui library was already there in 4.8, just not used yet?

**Comment by Bloke** [comment]
Nope. 4.9 only.

**Comment by Bloke** [comment]
Although there's nothing stopping us moving that forward.

**Comment by philwareham** [comment]
No it’s fine to leave til 4.9 then, I was mistaken.
